### PR TITLE
STL & PLY

### PR DESCRIPTION
**Pull request by AnastaZIuk**

### CODE REVIEW

**Comment by AnastaZIuk** [comment]
could you tell me what about [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/src/nbl/asset/interchange/CPLYMeshFileLoader.cpp#L302)? and should I precompute entire pipeline (all permutations) for **PLY** loader as well?

**Comment by devshgraphicsprogramming** [comment]
> 
> 
> could you tell me what about [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/src/nbl/asset/interchange/CPLYMeshFileLoader.cpp#L302)? and should I precompute entire pipeline (all permutations) for **PLY** loader as well?

someone made a `hash` parameter for the PLY loader's pipelines, but always set it to 1 or nothing (I dont rmemember) and I didnt get why the hash is needed as the PLY shaders never change (no compile-time options like UV vs no UV)

yes please precompute, I think there's only 3 permutations according to this:
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/src/nbl/asset/interchange/CPLYMeshFileLoader.cpp#L606

File: src/nbl/asset/interchange/CPLYMeshFileLoader.cpp [hunk 1]
```diff
@@ -22,7 +22,12 @@ namespace nbl
 namespace asset
 {
 
-CPLYMeshFileLoader::CPLYMeshFileLoader(IAssetManager* _am) : IRenderpassIndependentPipelineLoader(_am) {}
+CPLYMeshFileLoader::CPLYMeshFileLoader(IAssetManager* _am) 
+	: IRenderpassIndependentPipelineLoader(_am)
+{
+	initialize();
+	IRenderpassIndependentPipelineLoader::initialize();
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshFileLoader.cpp:29]: dont put initialize in the constructor

File: src/nbl/asset/interchange/CPLYMeshFileLoader.cpp [hunk 2]
```diff
@@ -57,6 +62,111 @@ bool CPLYMeshFileLoader::isALoadableFileFormat(io::IReadFile* _file) const
     return false;
 }
 
+void CPLYMeshFileLoader::initialize()
+{
+	auto precomputeAndCachePipeline = [&](CPLYMeshFileLoader::E_TYPE type, bool indexBufferBindingAvailable)
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshFileLoader.cpp:67]: put `IRenderpassIndependentPipelineLoader::initialize();` as the first thing this method does

Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshFileLoader.cpp:68]: done

File: src/nbl/asset/interchange/CPLYMeshFileLoader.cpp (hunk 1 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshFileLoader.cpp]: done

**Comment by AnastaZIuk** [comment]
I will replace std::vector attributes to SBufferBinding to avoid memcpy

File: src/nbl/asset/interchange/CSTLMeshFileLoader.cpp [hunk 3]
```diff
@@ -253,14 +254,18 @@ SAssetBundle CSTLMeshFileLoader::loadAsset(IReadFile* _file, const IAssetLoader:
...
+	quant_normal_t normal;
 	for (size_t i = 0u; i < positions.size(); ++i)
 	{
 		if (i % 3 == 0)
 			normal = quantNormalCache->quantize<EF_A2B10G10R10_SNORM_PACK32>(normals[i / 3]);
 		uint8_t* ptr = ((uint8_t*)(vertexBuf->getPointer())) + i * vtxSize;
 		memcpy(ptr, positions[i].pointer, 3 * 4);
-		((uint32_t*)(ptr + 12))[0] = normal;
+
+		((quant_normal_t*)((uint32_t*)(ptr + 12)))[0] = normal;
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:267]: no need for double cast, just do `*reinterpret_cast<quant_normal_t*>(ptr+12) = normal;`

Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:267]: `ptr` is `uint8_t*`, wrong pointer arithmetic will be applied

Comment (devshgraphicsprogramming) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:267]: here you're doing the exact same arithmetic.

you cast to `uint32_t` after incrementing the pointer by 12 bytes.

Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:267]: oh sorry of course

Comment (Crisspl) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:267]: also pls dont use c-style casts, it's already unreadable because of amount of parenthesis

Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp]: done

File: src/nbl/asset/interchange/CSTLMeshFileLoader.cpp [hunk 4]
```diff
@@ -24,6 +24,106 @@ constexpr auto COLOR_ATTRIBUTE = 1;
 constexpr auto UV_ATTRIBUTE = 2;
 constexpr auto NORMAL_ATTRIBUTE = 3;
 
+CSTLMeshFileLoader::CSTLMeshFileLoader(asset::IAssetManager* _m_assetMgr)
+	: IRenderpassIndependentPipelineLoader(_m_assetMgr), m_assetMgr(_m_assetMgr)
+{
+	initialize();
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:30]: stop calling `initialize` in the constructor

File: src/nbl/asset/interchange/CPLYMeshFileLoader.h [hunk 5]
```diff
@@ -54,6 +54,37 @@ class CPLYMeshFileLoader : public IRenderpassIndependentPipelineLoader
     virtual SAssetBundle loadAsset(io::IReadFile* _file, const IAssetLoader::SAssetLoadParams& _params, IAssetLoader::IAssetLoaderOverride* _override = nullptr, uint32_t _hierarchyLevel = 0u) override;
 private:
 
+	enum E_TYPE { ET_POS = 0, ET_UV = 2, ET_NORM = 3, ET_COL = 1 };
+
+	void initialize();
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshFileLoader.h:59]: use the `override` keyword always when you're overriding a virtual function

File: src/nbl/asset/interchange/CPLYMeshFileLoader.cpp [hunk 6]
```diff
@@ -24,7 +22,11 @@ namespace nbl
 namespace asset
 {
 
-CPLYMeshFileLoader::CPLYMeshFileLoader(IAssetManager* _am) : IRenderpassIndependentPipelineLoader(_am) {}
+CPLYMeshFileLoader::CPLYMeshFileLoader(IAssetManager* _am) 
+	: IRenderpassIndependentPipelineLoader(_am)
+{
+	initialize();
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshFileLoader.cpp:28]: sop calling `initialize` in the constructor

File: src/nbl/asset/IAssetManager.cpp [hunk 7]
```diff
@@ -394,10 +408,23 @@ void IAssetManager::insertBuiltinAssets()
...
     }
+
+    /*
+        Pipelines
+    */
+
+    //! Okay, an explaination is needed 
+    {
+
+    }
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/IAssetManager.cpp:429]: were you planning to do something here?

Comment (AnastaZIuk) [reply src/nbl/asset/IAssetManager.cpp:429]: oh it's irrelevant now

File: src/nbl/asset/interchange/CPLYMeshFileLoader.h (hunk 5 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshFileLoader.h:59]: I had no clue it's virtual

File: src/nbl/asset/IAssetManager.cpp (hunk 7 above)
Comment (devshgraphicsprogramming) [reply src/nbl/asset/IAssetManager.cpp:429]: actually what the STL and PLY loaders do in initialize coudl go here, cause PLY's materials cover all of STL's

Comment (AnastaZIuk) [reply src/nbl/asset/IAssetManager.cpp:429]: it was my first idea actually, but then plan then changed. Okay I will move it

File: src/nbl/asset/interchange/CPLYMeshWriter.cpp [hunk 8]
```diff
@@ -78,44 +78,81 @@ bool CPLYMeshWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
...
-	if (!file || !mesh || meshbuffers.size() > 1)
+	if (!file || !mesh)
 		return false;
+    
+    if (meshbuffers.size() > 1)
+    {
+        #ifdef  _NBL_DEBUG
+        os::Printer::log("PLY WRITER ERROR (" + std::to_string(__LINE__) + " line): Only one meshbuffer input is allowed for writing!", ELL_ERROR);
+        #endif // _NBL_DEBUG
+        return false;
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshWriter.cpp:89]: dont fail, just only write the first one

File: src/nbl/asset/interchange/CPLYMeshWriter.cpp [hunk 9]
```diff
@@ -78,44 +78,81 @@ bool CPLYMeshWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams&
...
+        const asset::E_FORMAT t = rawCopyMeshBuffer->getAttribFormat(0);
         std::string typeStr = getTypeString(t);
         vaidToWrite[0] = true;
         header +=
             "property " + typeStr + " x\n" +
             "property " + typeStr + " y\n" +
             "property " + typeStr + " z\n";
     }
-    if (meshBuffer->getAttribBoundBuffer(1).buffer)
+    if (rawCopyMeshBuffer->getAttribBoundBuffer(1).buffer)
```
Comment (devshgraphicsprogramming) [inline src/nbl/asset/interchange/CPLYMeshWriter.cpp:153]: wtf why are integer literals being used here?

ICPUMeshBuffer reports the attribute IDs used for position and normals, so use these.

For UV and Color you can assume 2 and 1 but make some `constexpr` with a proper name

Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshWriter.cpp:153]: todonigga

File: src/nbl/asset/interchange/CSTLMeshFileLoader.cpp (hunk 4 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CSTLMeshFileLoader.cpp:30]: corrected in following commit

File: src/nbl/asset/interchange/CPLYMeshFileLoader.h (hunk 5 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshFileLoader.h:59]: corrected in following commit

File: src/nbl/asset/interchange/CPLYMeshFileLoader.cpp (hunk 6 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshFileLoader.cpp:28]: corrected in following commit

File: src/nbl/asset/IAssetManager.cpp (hunk 7 above)
Comment (AnastaZIuk) [reply src/nbl/asset/IAssetManager.cpp:429]: irrelevant and solved

File: src/nbl/asset/interchange/CPLYMeshWriter.cpp (hunk 8 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshWriter.cpp:89]: corrected in following commit

File: src/nbl/asset/interchange/CPLYMeshWriter.cpp (hunk 9 above)
Comment (AnastaZIuk) [reply src/nbl/asset/interchange/CPLYMeshWriter.cpp:153]: corrected in following commit

**Pull request merged**